### PR TITLE
AP_Vehicle: add AIS

### PR DESCRIPTION
**Pull request by IamPete1**
Follow up to https://github.com/ArduPilot/ardupilot/pull/13839 moving into AP_Vehicle, still only enabled for rover by default but can now be enabled on other vehicles with the define.

### CODE REVIEW

**Comment by IamPete1** [comment]
I'm not sure what has gone wrong with semaphore, its passing that check locally.

**Comment by rmackay9** [comment]
@IamPete1,

Semaphore looks like it's complaining about this, "libraries/AP_Vehicle/AP_Vehicle.h: library header uses vehicle-dependent macros".  I think this is a real limitation we've added so either we need to remove that check or fix it

**Comment by IamPete1** [comment]
rebased to see if it will get past semaphore.

File: libraries/AP_Vehicle/AP_Vehicle.cpp [hunk 1]
```diff
@@ -67,6 +67,12 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
     AP_SUBGROUPINFO(efi, "EFI", 9, AP_Vehicle, AP_EFI),
 #endif
 
+#if !HAL_MINIMIZE_FEATURES && APM_BUILD_TYPE(APM_BUILD_Rover) // this should really be HAL_AIS_ENABLED
+    // @Group: AIS_
+    // @Path: ../AP_AIS/AP_AIS.cpp
+    AP_SUBGROUPINFO(ais, "AIS_",  10, AP_Vehicle, AP_AIS),
+#endif
```
Comment (IamPete1) [inline libraries/AP_Vehicle/AP_Vehicle.cpp:74]: @peterbarker @hendjoshsr71 I have reworked this to the dummy methods approach as in airspeed, any thoughts about how to get round the param table? Everything else works, not sure the flash cost yet......

Comment (IamPete1) [reply libraries/AP_Vehicle/AP_Vehicle.cpp:74]: We could change to a pointer table of course and load at init, but that seems overly complex.

**Comment by IamPete1** [comment]
The dummy methods cost us a bit on 2M boards,
```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  --------------  -----------  -------------  ----------------------------  -------------------------
arduplane        172 (+0.0098%)  0 (0.0000%)  4 (+0.0015%)   172 (+0.0098%)                                   206080
ardurover        -8 (-0.0005%)   0 (0.0000%)  0 (0.0000%)    -8 (-0.0005%)                                    363032
arducopter-heli  356 (+0.0200%)  0 (0.0000%)  -4 (-0.0015%)  356 (+0.0199%)                                   180144
arducopter       388 (+0.0219%)  0 (0.0000%)  4 (+0.0015%)   388 (+0.0218%)                                   187976
antennatracker   300 (+0.0221%)  0 (0.0000%)  4 (+0.0015%)   300 (+0.0220%)                                   604096
blimp            244 (+0.0188%)  0 (0.0000%)  4 (+0.0015%)   244 (+0.0187%)                                   663332
ardusub          204 (+0.0131%)  0 (0.0000%)  4 (+0.0015%)   204 (+0.0130%)                                   401900
```

No flash cost on 1M boards.

```
Binary Name      Text [B]          Data [B]     BSS (B)       Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  ------------  ----------------------------  -------------------------
arduplane        0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                        9400
ardurover        -4444 (-0.5189%)  0 (0.0000%)  4 (+0.0031%)  -4444 (-0.5184%)                                 130264
arducopter-heli  0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                       288
arducopter       0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                       47528
antennatracker   0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                      319584
blimp            0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                      301620
ardusub          0 (0.0000%)       0 (0.0000%)  0 (0.0000%)   0 (0.0000%)                                      169268
```
This does remove on Rover 1M boards, but I think that's OK. Currently this does not do any param conversion for rover, we could add it, AIS has now been out in a stable, @rmackay9 what do you think?

**Comment by auturgy** [comment]
I think we should really have this in ArduSub by default, and selfishly I'd like it in copter and plane too, although probably as a compile option.
@peterbarker

File: Tools/scripts/build_options.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self,
     Feature('Ident', 'ADSB_SAGETECH', 'HAL_ADSB_SAGETECH_ENABLED', 'Enable SageTech ADSB', 0, 'ADSB'),
     Feature('Ident', 'ADSB_SAGETECH_MXS', 'HAL_ADSB_SAGETECH_MXS_ENABLED', 'Enable SageTech MXS ADSB', 0, 'ADSB'),
     Feature('Ident', 'ADSB_UAVIONIX', 'HAL_ADSB_UAVIONIX_MAVLINK_ENABLED', 'Enable Uavionix ADSB', 0, 'ADSB'),
-    Feature('Ident', 'AIS', 'HAL_AIS_ENABLED', 'Enable AIS', 0, None),
+    Feature('Ident', 'AIS', 'AP_AIS_ENABLED', 'Enable AIS', 0, None),
```
Comment (tridge) [inline Tools/scripts/build_options.py:49]: maybe make AP_AIS_ENABLED=1 enable the full feature, and =2 enable the dummy only, and then in the header file default to dummy

**Review by tridge — CHANGES_REQUESTED** [review]
this will need wiki work to get the AIS docs in copter and plane, with a note on how to enable it with custom build server

**Comment by tridge** [comment]
@Hwurzburg note the doc changes needed to make AIS docs in copter and plane with note on how to enable in custom build server

File: libraries/AP_AIS/AP_AIS.cpp [hunk 3]
```diff
@@ -804,4 +821,26 @@ bool AP_AIS::decode_latest_term()
...
 }
 
-#endif  // HAL_AIS_ENABLED
+// get singleton instance
+AP_AIS *AP_AIS::get_singleton() {
+    return _singleton;
+}
+
+#else
+
```
Comment (tridge) [inline libraries/AP_AIS/AP_AIS.cpp:830]: add comment explaining this dummy code (why its needed)

**Review by peterbarker — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**